conbench/api/compare.py

```diff
@@ -1,146 +1,275 @@
-        # The goal here was probably to raise ValueError?
-        # try:
-        #     baseline_id, contender_id = compare_ids.split("...", 1)
-        # except ValueError:
-        #     self.abort_404_not_found()
+    class AugmentedBenchmarkResult(BenchmarkResult):
+        """For type hints only. This represents a BenchmarkResult that has been
+        augmented in-place with the following augmentation functions, as is done in this
+        module:
 
-        return threshold, threshold_z
+        - set_z_scores()
+        - set_display_benchmark_name()
+        - set_display_case_permutation()
 
+        TODO: remove this and replace with actual BenchmarkResult properties.
+        """
 
-class CompareEntityEndpoint(ApiEndpoint, CompareMixin):
-    def _get_results(self, rid: str) -> List[BenchmarkResult]:
-        raise NotImplementedError
+        display_bmname: str
+        display_case_perm: str
+        z_score: Optional[float]
+
+
+class UnmatchingUnitsError(Exception):
+    pass
+
+
+class BenchmarkResultComparator:
+    """Data model class to hold the comparison of two BenchmarkResults."""
+
+    def __init__(
+        self,
+        baseline: Optional["AugmentedBenchmarkResult"],
+        contender: Optional["AugmentedBenchmarkResult"],
+        threshold: Optional[float],
+        threshold_z: Optional[float],
+    ) -> None:
+        if (
+            baseline
+            and baseline.unit
+            and contender
+            and contender.unit
+            and baseline.unit != contender.unit
+        ):
+            raise UnmatchingUnitsError(
+                "Benchmark units do not match. Benchmark result with ID "
+                f"'{baseline.id}' has unit '{baseline.unit}' and benchmark result "
+                f"with ID '{contender.id}' has unit '{contender.unit}'."
+            )
+
+        self.baseline = baseline
+        self.contender = contender
+        self.threshold = (
+            float(threshold)
+            if threshold is not None
+            else DEFAULT_PAIRWISE_PERCENT_THRESHOLD
+        )
+        self.threshold_z = (
+            float(threshold_z) if threshold_z is not None else DEFAULT_Z_SCORE_THRESHOLD
+        )
+
+    @property
+    def unit(self) -> str:
+        if self.baseline and self.baseline.unit:
+            return self.baseline.unit
+        if self.contender and self.contender.unit:
+            return self.contender.unit
+        return "unknown"
+
+    @property
+    def less_is_better(self) -> bool:
+        return _less_is_better(self.unit)
+
+    @staticmethod
+    def result_info(result: Optional["AugmentedBenchmarkResult"]) -> Optional[dict]:
+        if not result:
+            return None
+
+        return {
+            "benchmark_result_id": result.id,
+            "benchmark_name": result.display_bmname,
+            "case_permutation": result.display_case_perm,
+            "language": result.context.tags.get("benchmark_language", "unknown"),
```

Comment:
I think it's validated somewhere (well, by using it) that this has to be populated; results without it fail to post. But the defensiveness is good in case that ever gets relaxed